homeassistant/components/amcrest/camera.py

```diff
@@ -43,12 +47,18 @@ def __init__(self, hass, amcrest):
         self._stream_source = amcrest.stream_source
         self._resolution = amcrest.resolution
         self._token = self._auth = amcrest.authentication
+        self._snapshot_lock = threading.Lock()
 
     def camera_image(self):
         """Return a still image response from the camera."""
-        # Send the request to snap a picture and return raw jpg data
-        response = self._camera.snapshot(channel=self._resolution)
-        return response.data
+        with self._snapshot_lock:
```

Comment:
We should avoid waiting on locks in our thread pool. We can switch to the async api and use an `asyncio.Lock` instead.

Reply:
Oh. I agree that would be better but are you implying that we actually have a rule against locks?

Reply:
Yeah, it's basically the same rule as for sleeping in the thread pool.

Reply:
Hm. I think that dilutes the value of the thread pool. What if the lock was in the third party library? Maybe we should disallow I/O as well since it will sleep?

Reply:
@pnbruckner I think this is the rewrite that Martin suggests:
```python
    async def async_camera_image(self):
        """Return a still image response from the camera."""
        with self._snapshot_lock:
            try:
                # Send the request to snap a picture and return raw jpg data
                response = await hass.async_add_executor_job(
                    self._camera.snapshot, self._resolution)
                return response.data
            except (RequestException, ReadTimeoutError, ValueError) as error:
                _LOGGER.error(
                    'Could not get camera image due to error %s', error)
                return None
```
(and then `_snapshot_lock` should be an `asyncio.Lock()`)

Reply:
Locks in third party libraries that are active in our thread pool are equally bad. I try to inform about that whenever I catch one. Super slow I/O should also be preferably be done in a separate single thread. Hopefully the user don't use too many of those together.

We have a good alternative by switching to async api.

Reply:
Ok. So how do I go about this? Do I restore my branch and check in more changes? Do I open a new PR? Something else?

Reply:
That should be a new PR.

Reply:
@amelchio @MartinHjelmare

I need some help on this. This is what I came up with based on the suggested code above:
```python
        self._snapshot_lock = asyncio.Lock()

    async def async_camera_image(self):
        """Return a still image response from the camera."""
        with await self._snapshot_lock:
            try:
                # Send the request to snap a picture and return raw jpg data
                _LOGGER.critical('async_camera_image start')
                response = await self.hass.async_add_executor_job(
                    self._camera.snapshot, self._resolution)
                _LOGGER.critical('async_camera_image end: %i', len(response.data))
                return response.data
            except (RequestException, ReadTimeoutError, ValueError) as error:
                _LOGGER.error(
                    'Could not get camera image due to error %s', error)
                return None
```
But it doesn't seem to be working. I added the logging temporarily to try and see what's going on. I also added logging to the self._camera.snapshot method. Here is what I'm seeing:
```text
2019-03-05 17:53:52 CRITICAL (MainThread) [homeassistant.components.amcrest.camera] async_camera_image start
2019-03-05 17:53:52 CRITICAL (SyncWorker_1) [amcrest.snapshot] snapshot start
2019-03-05 17:53:53 CRITICAL (SyncWorker_1) [amcrest.snapshot] snapshot end
2019-03-05 17:54:00 CRITICAL (MainThread) [homeassistant.components.amcrest.camera] async_camera_image end: 158840
2019-03-05 17:54:00 CRITICAL (MainThread) [homeassistant.components.amcrest.camera] async_camera_image start
2019-03-05 17:54:00 CRITICAL (SyncWorker_7) [amcrest.snapshot] snapshot start
2019-03-05 17:54:01 CRITICAL (SyncWorker_7) [amcrest.snapshot] snapshot end
2019-03-05 17:54:09 CRITICAL (MainThread) [homeassistant.components.amcrest.camera] async_camera_image end: 159495
2019-03-05 17:54:09 CRITICAL (MainThread) [homeassistant.components.amcrest.camera] async_camera_image start
2019-03-05 17:54:09 CRITICAL (SyncWorker_2) [amcrest.snapshot] snapshot start
2019-03-05 17:54:09 CRITICAL (MainThread) [homeassistant.components.amcrest.camera] async_camera_image start
2019-03-05 17:54:09 CRITICAL (SyncWorker_8) [amcrest.snapshot] snapshot start
2019-03-05 17:54:09 WARNING (SyncWorker_2) [urllib3.connectionpool] Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ProtocolError('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))': /cgi-bin/snapshot.cgi?channel=0
2019-03-05 17:54:09 WARNING (SyncWorker_2) [amcrest.http] Trying again due to error 401 Client Error: Unauthorized for url: http://192.168.1.164:10007/cgi-bin/snapshot.cgi?channel=0
```
This shows two successful snapshots, but even so, there is a long delay from the snapshot method finishing to async_camera_image finishing. I don't understand that. Then you can see async_camera_image starts twice, which results in two calls to snapshot starting twice, which leads to the errors that the locking is supposed to be avoiding.

Why does the locking seem like it's not working?

Reply:
> Why does the locking seem like it's not working?

Sorry, try this:
```python
async with self._snapshot_lock:
```
... but please do open a new PR (you can tag me).

Reply:
Thank you. That fixed it! (And now I see why. Still learning asyncio!)

I'll be submitting a new PR with the change soon.